src/hotspot/share/opto/castnode.cpp

```diff
@@ -213,13 +213,7 @@ const Type* CastIINode::Value(PhaseGVN* phase) const {
 
   // Similar to ConvI2LNode::Value() for the same reasons
   // see if we can remove type assertion after loop opts
-  // But here we have to pay extra attention:
-  // Do not narrow the type of range check dependent CastIINodes to
-  // avoid corruption of the graph if a CastII is replaced by TOP but
-  // the corresponding range check is not removed.
-  if (!_range_check_dependency) {
-    res = widen_type(phase, res, T_INT);
-  }
```

Comment:
Can you explain why you changed this, and why it is ok?

Reply:
`ConvI2L` has a similar transformation. Let's say we have 2 `ConvI2L` nodes with identical inputs but different types:
```
(ConvI2L _ input [2..max_jint])
(ConvI2L _ input [1..max_jint])
```
They are transformed to:
```
(ConvI2L _ input [0..max_jint])
(ConvI2L _ input [0..max_jint])
```
so they can common. With range checks, the pattern is:
```
(ConvI2L _ (CastII control input [2..max_jint]) [2..max_jint])
(ConvI2L _ (CastII control input [1..max_jint]) [1..max_jint]
```
Without this patch, the range checks `CastII` are removed after loop opts so having the transformation be done only on `ConvI2L` nodes is sufficient. With this change the `CastII` are left in the IR so they need to be transformed the same way:

```
(ConvI2L _ (CastII control input [0..max_jint]) [0..max_jint])
(ConvI2L _ (CastII control input [0..max_jint]) [0..max_jint]
```

so they can common and the `ConvI2L` then common.